### ISSUE TITLE
CloudFormation - Sub Intrinsic Function doesn't support variable map

### ISSUE THREAD

**Issue by Tobolov** [localstack/localstack#3333]
# Type of request: This is a ...

[ X] bug report
[ ] feature request

# Detailed description
The current !Sub implementation only supports `!Sub String`. It doesn't support:
```yaml
!Sub
  - String
  - Var1Name: Var1Value
    Var2Name: Var2Value
```

Details on how the sub intrinsic function behaves when given a variable map can be seen here:  [Intrinsic Function Reference Sub](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-sub.html)

## Expected behaviour
The sub intrinsic function supports being provided a variable map.

## Actual behaviour
The sub intrinsic function supports doesn't support being provided a variable map. The prevents resources which use this intrinsic function to not be created (e.g. in the below template, the SnsTopic resource is not created). The error outputted by localstack in debug mode is:
```
2020-12-09T13:44:16:ERROR:localstack.services.cloudformation.cloudformation_starter: Unable to parse and create resource "SnsTopic": replace() argument 2 must be str, not dict Traceback (most recent call last):
  File "/home/tobolov/linrepos/localstack/localstack/services/cloudformation/cloudformation_starter.py", line 365, in parse_and_create_resource
    result = _parse_and_create_resource(
  File "/home/tobolov/linrepos/localstack/localstack/services/cloudformation/cloudformation_starter.py", line 429, in _parse_and_create_resource
    resource_tuple = parsing.parse_resource_and_generate_name(logical_id, resource_json, resources_map)
  File "/home/tobolov/linrepos/localstack/.venv/lib/python3.8/site-packages/moto/cloudformation/parsing.py", line 282, in parse_resource_and_generate_name
    resource_tuple = parse_resource(resource_json, resources_map)
  File "/home/tobolov/linrepos/localstack/.venv/lib/python3.8/site-packages/moto/cloudformation/parsing.py", line 274, in parse_resource
    resource_json = clean_json(resource_json, resources_map)
  File "/home/tobolov/linrepos/localstack/localstack/services/cloudformation/cloudformation_starter.py", line 290, in clean_json
    rs = clean_json_orig(resource_json, resources_map)
  File "/home/tobolov/linrepos/localstack/.venv/lib/python3.8/site-packages/moto/cloudformation/parsing.py", line 202, in clean_json
    cleaned_val = clean_json(value, resources_map)
  File "/home/tobolov/linrepos/localstack/localstack/services/cloudformation/cloudformation_starter.py", line 290, in clean_json
    rs = clean_json_orig(resource_json, resources_map)
  File "/home/tobolov/linrepos/localstack/.venv/lib/python3.8/site-packages/moto/cloudformation/parsing.py", line 202, in clean_json
    cleaned_val = clean_json(value, resources_map)
  File "/home/tobolov/linrepos/localstack/localstack/services/cloudformation/cloudformation_starter.py", line 340, in clean_json
    resource_json['Fn::Sub'][0] = resource_json['Fn::Sub'][0].replace('${%s}' % k, v)
TypeError: replace() argument 2 must be str, not dict
```

# Steps to reproduce
1. Deploy the following CloudFormation template:
```yaml
---
AWSTemplateFormatVersion: "2010-09-09"
Parameters:
  Environment:
    Type: String
    Default: 'companyname-ci'
Resources:
  SnsTopic:
    Type: AWS::SNS::Topic
    Properties: 
      TopicName: !Sub 
        - '${Env}-slack-topic'
        - { Env: !Select [1, !Split ["-" , !Ref Environment]] }
```
2. Run `aws sns list-topics --endpoint-url http://localhost:4566 --region us-east-1`
3. Observe that no topics exist, the topic listed in the CloudFormation template was not created.

**Comment by whummer** [comment]
Thanks for reporting @Tobolov . This should be fixed in #3470 - can you please pull the `latest` Docker image and give it another try? Thanks

**Comment by AffiTheCreator** [comment]
Hello @whummer and @Tobolov 

I'm experiencing the same result has described in the issue. I'm currently trying to deploy an [aws-sam app](https://github.com/aws-samples/aws-marketplace-serverless-saas-integration). 

I'm trying to use localstack to test a lambda I have modified to add some functionality to the metering section. 

I managed to deploy the stack via samcli commands into AWS, the problem is when I run it against the localstack framework it fails to create the HourlyRole with ```ResourceStatusReason: replace() argument 2 must be str, not dict```

below is the script to create the necessary resources before deploying the stack

```bash
    # Create SNS Topic for subscription
    sns_topic_arn=$(awslocal sns create-topic --name subscriptions-marketplace-saas-dev | jq -r '.TopicArn')

    # aws --endpoint-url=http://localhost:4566 sns create-topic --name subscriptions-marketplace-saas-dev
    subscription_sns_topic=$sns_topic_arn

    # Create the S3 bucket
    s3_bucket_name_saas="seller-saas-dev"
    # aws s3api create-bucket --bucket $s3_bucket_name_saas
    awslocal s3api create-bucket --bucket $s3_bucket_name_saas

    s3_bucket_name="seller-cf-dev"
    awslocal s3api create-bucket --bucket $s3_bucket_name

```

Trust-policy.json

```json
{
    "Version": "2012-10-17",
    "Statement": [
      {
        "Effect": "Allow",
        "Principal": {
          "Service": "cloudformation.amazonaws.com"
        },
        "Action": "sts:AssumeRole"
      }
    ]
  }
```

To facilitate, here is the snippet of the template.yaml that is responsible for building the Hourly lambda, since its the HourlyRole that is failing, i assume the problem is here. 

```yaml

  Hourly:
    Type: AWS::Serverless::Function
    Condition: CreateSubscriptionLogic
    Properties:
      CodeUri: src
      Handler: metering-hourly-job.job
      Runtime: nodejs18.x
      Environment:
        Variables:
          SQSMeteringRecordsUrl: !Ref SQSMeteringRecords
          AWSMarketplaceMeteringRecordsTableName: !Ref AWSMarketplaceMeteringRecordsTableName
      Policies:
        - DynamoDBReadPolicy:
            TableName: !Ref AWSMarketplaceMeteringRecordsTableName
        - SQSSendMessagePolicy:
            QueueName: !GetAtt SQSMeteringRecords.QueueName
      Events:
        CWSchedule:
          Type: Schedule
          Properties:
            Schedule: "rate(1 hour)"
            Name: !Join ["-", [MeteringSchedule, !Ref AWS::StackName]]
            Description: SaaS Metering
            Enabled: TRUE
```

Below is my docker compose file 

```yaml
version: "3.8"

services:
  localstack:
    container_name: "${LOCALSTACK_DOCKER_NAME-localstack_main}"
    image: localstack/localstack-pro
    ports:
      - "127.0.0.1:4566:4566"            # LocalStack Gateway
      - "127.0.0.1:4510-4559:4510-4559"  # ext services port range
      - "127.0.0.1:53:53"                # DNS config
      - "127.0.0.1:53:53/udp"            # DNS config
      - "127.0.0.1:443:443"              # LocalStack HTTPS Gateway
    environment:
      - DEBUG=1
      - PERSISTENCE=${PERSISTENCE-}
      - LAMBDA_EXECUTOR=${LAMBDA_EXECUTOR-}
      - LOCALSTACK_API_KEY="XXXXXXXXX"
      - DOCKER_HOST=unix:///var/run/docker.sock   
    volumes:
      - "${LOCALSTACK_VOLUME_DIR:-./volume}:/var/lib/localstack"
      - "/var/run/docker.sock:/var/run/docker.sock"
```
Below is my stack trace

```terminal

        Deploying with following values
        ===============================
        Stack name                   : marketplace-saas-dev
        Region                       : us-east-1
        Confirm changeset            : False
        Disable rollback             : False
        Deployment s3 bucket         : None
        Capabilities                 : ["CAPABILITY_IAM"]
        Parameter overrides          : {"WebsiteS3BucketName": "seller-saas-dev", "ProductCode": "xxxxxxxxxxxxxxxxxxxxxxxxxxx", "EntitlementSNSTopic": "arn:aws:sns:us-east-1:000000000000:subscriptions-marketplace-saas-dev", "SubscriptionSNSTopic": "arn:aws:sns:us-east-1:000000000000:subscriptions-marketplace-saas-dev", "MarketplaceTechAdminEmail": "dev.seller@example.ai", "TypeOfSaaSListing": "subscriptions"}
        Signing Profiles             : {}

Initiating deployment
=====================

2023-07-24 22:06:23,544 | Collected default values for parameters: {'WebsiteS3BucketName': '', 'NewSubscribersTableName': 'AWSMarketplaceSubscribers', 'AWSMarketplaceMeteringRecordsTableName':
'AWSMarketplaceMeteringRecords', 'TypeOfSaaSListing': 'subscriptions', 'MarketplaceSellerEmail': '', 'EntitlementSNSTopic': '', 'SubscriptionSNSTopic': '', 'CreateRegistrationWebPage': True}
2023-07-24 22:06:23,555 | Sam customer defined id is more priority than other IDs. Customer defined id for resource CloudFrontOriginAccessIdentity is CloudFrontOriginAccessIdentity
2023-07-24 22:06:23,556 | Sam customer defined id is more priority than other IDs. Customer defined id for resource CloudfrontDistribution is CloudfrontDistribution
2023-07-24 22:06:23,556 | Sam customer defined id is more priority than other IDs. Customer defined id for resource WebsiteS3Bucket is WebsiteS3Bucket
2023-07-24 22:06:23,557 | Sam customer defined id is more priority than other IDs. Customer defined id for resource WebsiteS3BucketLog is WebsiteS3BucketLog
2023-07-24 22:06:23,558 | Sam customer defined id is more priority than other IDs. Customer defined id for resource S3BucketPolicy is S3BucketPolicy
2023-07-24 22:06:23,558 | Sam customer defined id is more priority than other IDs. Customer defined id for resource LambdaEdgeRedirectPostRequests is LambdaEdgeRedirectPostRequests
2023-07-24 22:06:23,559 | Sam customer defined id is more priority than other IDs. Customer defined id for resource AWSMarketplaceMeteringRecords is AWSMarketplaceMeteringRecords
2023-07-24 22:06:23,559 | Sam customer defined id is more priority than other IDs. Customer defined id for resource AWSMarketplaceSubscribers is AWSMarketplaceSubscribers
2023-07-24 22:06:23,560 | Sam customer defined id is more priority than other IDs. Customer defined id for resource RegisterNewMarketplaceCustomer is RegisterNewMarketplaceCustomer
2023-07-24 22:06:23,560 | Sam customer defined id is more priority than other IDs. Customer defined id for resource EntitlementSQSQueue is EntitlementSQSQueue
2023-07-24 22:06:23,561 | Sam customer defined id is more priority than other IDs. Customer defined id for resource EntitlementSQSHandler is EntitlementSQSHandler
2023-07-24 22:06:23,561 | Sam customer defined id is more priority than other IDs. Customer defined id for resource SubscriptionSQSHandler is SubscriptionSQSHandler
2023-07-24 22:06:23,562 | Sam customer defined id is more priority than other IDs. Customer defined id for resource SupportSNSTopic is SupportSNSTopic
2023-07-24 22:06:23,562 | Sam customer defined id is more priority than other IDs. Customer defined id for resource GrantOrRevokeAccess is GrantOrRevokeAccess
2023-07-24 22:06:23,563 | Sam customer defined id is more priority than other IDs. Customer defined id for resource Hourly is Hourly
2023-07-24 22:06:23,563 | Sam customer defined id is more priority than other IDs. Customer defined id for resource SQSMeteringRecords is SQSMeteringRecords
2023-07-24 22:06:23,564 | Sam customer defined id is more priority than other IDs. Customer defined id for resource MeteringSQSHandler is MeteringSQSHandler
2023-07-24 22:06:23,564 | There is no customer defined id or cdk path defined for resource ServerlessRestApi, so we will use the resource logical id as the resource id
2023-07-24 22:06:23,566 | Unable to resolve property DomainName: OrderedDict([('Fn::GetAtt', ['WebsiteS3Bucket', 'DomainName'])]). Leaving as is.
2023-07-24 22:06:23,567 | Unable to resolve property Bucket: OrderedDict([('Fn::GetAtt', ['WebsiteS3BucketLog', 'DomainName'])]). Leaving as is.
2023-07-24 22:06:23,568 | Unable to resolve property MarketplaceSellerEmail: OrderedDict([('Fn::If', ['Buyernotificationemail', OrderedDict([('Ref', 'MarketplaceSellerEmail')]), OrderedDict([('Ref',
'AWS::NoValue')])])]). Leaving as is.
2023-07-24 22:06:23,569 | Unable to resolve property Policies: [{'Statement': [{'Action': ['dynamodb:PutItem', 'dynamodb:UpdateItem', 'dynamodb:BatchWriteItem'], 'Effect': 'Allow', 'Resource': [{'Fn::Sub':
['arn:${AWS::Partition}:dynamodb:${AWS::Region}:${AWS::AccountId}:table/${tableName}', {'tableName': OrderedDict([('Ref', 'NewSubscribersTableName')])}]}, {'Fn::Sub':
['arn:${AWS::Partition}:dynamodb:${AWS::Region}:${AWS::AccountId}:table/${tableName}/index/*', {'tableName': OrderedDict([('Ref', 'NewSubscribersTableName')])}]}]}]}, OrderedDict([('Statement',
[OrderedDict([('Sid', 'AWSMarketplaceResolveCustomer'), ('Effect', 'Allow'), ('Action', ['aws-marketplace:ResolveCustomer']), ('Resource', '*')])])]), OrderedDict([('Fn::If', ['CreateEntitlementLogic',
OrderedDict([('Statement', [OrderedDict([('Sid', 'SQSPolicy'), ('Effect', 'Allow'), ('Action', ['sqs:SendMessage']), ('Resource', OrderedDict([('Fn::GetAtt', ['EntitlementSQSQueue', 'Arn'])]))])])]),
OrderedDict([('Ref', 'AWS::NoValue')])])]), OrderedDict([('Statement', [OrderedDict([('Sid', 'Sendemail'), ('Effect', 'Allow'), ('Action', ['ses:SendEmail']), ('Resource', '*')])])])]. Leaving as is.
2023-07-24 22:06:23,572 | Unable to resolve property Stream: OrderedDict([('Fn::GetAtt', ['AWSMarketplaceSubscribers', 'StreamArn'])]). Leaving as is.
2023-07-24 22:06:23,573 | Unable to resolve property queueName: OrderedDict([('Fn::GetAtt', ['SQSMeteringRecords', 'QueueName'])]). Leaving as is.
2023-07-24 22:06:23,574 | Unable to resolve property Resource: {'Fn::Sub': ['arn:${AWS::Partition}:sqs:${AWS::Region}:${AWS::AccountId}:${queueName}', {'queueName': OrderedDict([('Fn::GetAtt',
['SQSMeteringRecords', 'QueueName'])])}]}. Leaving as is.
2023-07-24 22:06:23,575 | 0 stacks found in the template
2023-07-24 22:06:23,575 | Collected default values for parameters: {'WebsiteS3BucketName': '', 'NewSubscribersTableName': 'AWSMarketplaceSubscribers', 'AWSMarketplaceMeteringRecordsTableName':
'AWSMarketplaceMeteringRecords', 'TypeOfSaaSListing': 'subscriptions', 'MarketplaceSellerEmail': '', 'EntitlementSNSTopic': '', 'SubscriptionSNSTopic': '', 'CreateRegistrationWebPage': True}
2023-07-24 22:06:23,588 | Sam customer defined id is more priority than other IDs. Customer defined id for resource CloudFrontOriginAccessIdentity is CloudFrontOriginAccessIdentity
2023-07-24 22:06:23,589 | Sam customer defined id is more priority than other IDs. Customer defined id for resource CloudfrontDistribution is CloudfrontDistribution
2023-07-24 22:06:23,590 | Sam customer defined id is more priority than other IDs. Customer defined id for resource WebsiteS3Bucket is WebsiteS3Bucket
2023-07-24 22:06:23,590 | Sam customer defined id is more priority than other IDs. Customer defined id for resource WebsiteS3BucketLog is WebsiteS3BucketLog
2023-07-24 22:06:23,591 | Sam customer defined id is more priority than other IDs. Customer defined id for resource S3BucketPolicy is S3BucketPolicy
2023-07-24 22:06:23,591 | Sam customer defined id is more priority than other IDs. Customer defined id for resource LambdaEdgeRedirectPostRequests is LambdaEdgeRedirectPostRequests
2023-07-24 22:06:23,592 | Sam customer defined id is more priority than other IDs. Customer defined id for resource AWSMarketplaceMeteringRecords is AWSMarketplaceMeteringRecords
2023-07-24 22:06:23,593 | Sam customer defined id is more priority than other IDs. Customer defined id for resource AWSMarketplaceSubscribers is AWSMarketplaceSubscribers
2023-07-24 22:06:23,593 | Sam customer defined id is more priority than other IDs. Customer defined id for resource RegisterNewMarketplaceCustomer is RegisterNewMarketplaceCustomer
2023-07-24 22:06:23,594 | Sam customer defined id is more priority than other IDs. Customer defined id for resource EntitlementSQSQueue is EntitlementSQSQueue
2023-07-24 22:06:23,594 | Sam customer defined id is more priority than other IDs. Customer defined id for resource EntitlementSQSHandler is EntitlementSQSHandler
2023-07-24 22:06:23,595 | Sam customer defined id is more priority than other IDs. Customer defined id for resource SubscriptionSQSHandler is SubscriptionSQSHandler
2023-07-24 22:06:23,595 | Sam customer defined id is more priority than other IDs. Customer defined id for resource SupportSNSTopic is SupportSNSTopic
2023-07-24 22:06:23,596 | Sam customer defined id is more priority than other IDs. Customer defined id for resource GrantOrRevokeAccess is GrantOrRevokeAccess
2023-07-24 22:06:23,597 | Sam customer defined id is more priority than other IDs. Customer defined id for resource Hourly is Hourly
2023-07-24 22:06:23,597 | Sam customer defined id is more priority than other IDs. Customer defined id for resource SQSMeteringRecords is SQSMeteringRecords
2023-07-24 22:06:23,598 | Sam customer defined id is more priority than other IDs. Customer defined id for resource MeteringSQSHandler is MeteringSQSHandler
2023-07-24 22:06:23,598 | There is no customer defined id or cdk path defined for resource ServerlessRestApi, so we will use the resource logical id as the resource id
2023-07-24 22:06:23,600 | Unable to resolve property DomainName: OrderedDict([('Fn::GetAtt', ['WebsiteS3Bucket', 'DomainName'])]). Leaving as is.
2023-07-24 22:06:23,601 | Unable to resolve property Bucket: OrderedDict([('Fn::GetAtt', ['WebsiteS3BucketLog', 'DomainName'])]). Leaving as is.
2023-07-24 22:06:23,602 | Unable to resolve property MarketplaceSellerEmail: OrderedDict([('Fn::If', ['Buyernotificationemail', OrderedDict([('Ref', 'MarketplaceSellerEmail')]), OrderedDict([('Ref',
'AWS::NoValue')])])]). Leaving as is.
2023-07-24 22:06:23,603 | Unable to resolve property Policies: [{'Statement': [{'Action': ['dynamodb:PutItem', 'dynamodb:UpdateItem', 'dynamodb:BatchWriteItem'], 'Effect': 'Allow', 'Resource': [{'Fn::Sub':
['arn:${AWS::Partition}:dynamodb:${AWS::Region}:${AWS::AccountId}:table/${tableName}', {'tableName': OrderedDict([('Ref', 'NewSubscribersTableName')])}]}, {'Fn::Sub':
['arn:${AWS::Partition}:dynamodb:${AWS::Region}:${AWS::AccountId}:table/${tableName}/index/*', {'tableName': OrderedDict([('Ref', 'NewSubscribersTableName')])}]}]}]}, OrderedDict([('Statement',
[OrderedDict([('Sid', 'AWSMarketplaceResolveCustomer'), ('Effect', 'Allow'), ('Action', ['aws-marketplace:ResolveCustomer']), ('Resource', '*')])])]), OrderedDict([('Fn::If', ['CreateEntitlementLogic',
OrderedDict([('Statement', [OrderedDict([('Sid', 'SQSPolicy'), ('Effect', 'Allow'), ('Action', ['sqs:SendMessage']), ('Resource', OrderedDict([('Fn::GetAtt', ['EntitlementSQSQueue', 'Arn'])]))])])]),
OrderedDict([('Ref', 'AWS::NoValue')])])]), OrderedDict([('Statement', [OrderedDict([('Sid', 'Sendemail'), ('Effect', 'Allow'), ('Action', ['ses:SendEmail']), ('Resource', '*')])])])]. Leaving as is.
2023-07-24 22:06:23,606 | Unable to resolve property Stream: OrderedDict([('Fn::GetAtt', ['AWSMarketplaceSubscribers', 'StreamArn'])]). Leaving as is.
2023-07-24 22:06:23,607 | Unable to resolve property queueName: OrderedDict([('Fn::GetAtt', ['SQSMeteringRecords', 'QueueName'])]). Leaving as is.
2023-07-24 22:06:23,607 | Unable to resolve property Resource: {'Fn::Sub': ['arn:${AWS::Partition}:sqs:${AWS::Region}:${AWS::AccountId}:${queueName}', {'queueName': OrderedDict([('Fn::GetAtt',
['SQSMeteringRecords', 'QueueName'])])}]}. Leaving as is.
2023-07-24 22:06:23,609 | 18 resources found in the stack


Waiting for changeset to be created..

CloudFormation stack changeset
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Operation                                           LogicalResourceId                                   ResourceType                                        Replacement
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
* Modify                                            SubscriptionSQSHandlerRole                          AWS::IAM::Role                                      False
* Modify                                            CloudfrontDistribution                              AWS::CloudFront::Distribution                       False
* Modify                                            HourlyRole                                          AWS::IAM::Role                                      False
* Modify                                            SupportSNSTopic                                     AWS::SNS::Topic                                     False
* Modify                                            ServerlessRestApi                                   AWS::ApiGateway::RestApi                            False
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------


Changeset created successfully. arn:aws:cloudformation:us-east-1:000000000000:changeSet/samcli-deploy1690232783/eb277070


2023-07-24 22:06:23 - Waiting for stack create/update to complete

CloudFormation events from stack operations (refresh every 5.0 seconds)
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
ResourceStatus                                      ResourceType                                        LogicalResourceId                                   ResourceStatusReason
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
UPDATE_COMPLETE                                     AWS::SQS::QueuePolicy                               EntitlementSQSHandlerMySQSEventQueuePolicy          -
UPDATE_COMPLETE                                     AWS::Lambda::EventSourceMapping                     EntitlementSQSHandlerMySQSEventEventSourceMapping   -
UPDATE_COMPLETE                                     AWS::SNS::Subscription                              EntitlementSQSHandlerMySQSEvent                     -
UPDATE_FAILED                                       AWS::IAM::Role                                      HourlyRole                                          replace() argument 2 must be str, not dict
```

Thank you so much for the app so far, im running some trial with pro version because i wanted to test the security to , but maybe i need some guidance. I have tried your docs , ive tried google but could not find anything related to localstack + aws-marketplace-serverless-saas-integration
